File: bench/06_newton_sqrt/bench.py
"""Benchmark 6: Newton's method square root (100,000 calls of 50 steps).

The timed loop runs `--n` ops, each on an input that differs by the repetition number in
VALUE, never in the amount of work, and folds every result into `sink`.  The `hash`
column is one canonical op's result; every lane must print the same one.
"""
import sys
import time

import math


def newton_sqrt(x):
    guess = x / 2.0
    for _ in range(50):
        guess = (guess + x / guess) / 2.0
    return guess


def roots(calls, salt):
    acc = 0.0
    for j in range(calls):
        acc += newton_sqrt(float(j + 1 + salt))
    return int(math.floor(acc + 0.5))

def arg_n(dflt):
    """`--n N`: how many ops the timed loop runs (the harness calibrates it per lane)."""
    n = dflt
    for i, a in enumerate(sys.argv):
        if a == "--n" and i + 1 < len(sys.argv):
            n = int(sys.argv[i + 1])
    return max(n, 2)


def row(name, iters, us, items, result, sink):
    """One measured routine, in the row format every lane prints (bench/README.md)."""
    print("routine\titers\tus\tns_op\tpx\tns_px\thash")
    print(f"{name}\t{iters}\t{us}\t{us * 1000 // iters}\t{items}\t{us * 1000 / (iters * items):.3f}\t{result:x}")
    print(f"time: {us // 1000}ms sink={sink}")


def timed(n, op):
    t0 = time.perf_counter_ns()
    sink = 0
    for r in range(n):
        sink += op(r)
    return (time.perf_counter_ns() - t0) // 1000, sink

if __name__ == "__main__":
    n = arg_n(2)
    calls = 100_000
    us, sink = timed(n, lambda r: roots(calls, r & 1))
    row("newton_sqrt", n, us, calls, roots(calls, 0), sink)

File: bench/02_sum_loop/bench.rs
// Benchmark 2: Integer loop (5 million steps of a loop-carried mix)
//
// One OP is 5,000,000 steps of `acc = ((acc ^ i) + (i & 7)) & 0xFFFFFF` — loop-carried, so it has no
// closed form and cannot be vectorised: the row measures the scalar loop itself.  The timed loop runs `--n` ops, each on an input that differs by
// the repetition number — in VALUE, never in the amount of work — so no optimiser can
// hoist or collapse the kernel across repetitions, and folds every result into `sink`.
// The `hash` column is one canonical op's result: every lane must print the same one,
// or the lanes are not computing the same thing and their times do not compare.
use std::hint::black_box;
use std::time::Instant;

fn mix(count: i64, salt: i64) -> i64 {
    let mut acc = salt;
    for i in 0..count {
        acc = ((acc ^ i) + (i & 7)) & 16777215;
    }
    acc
}

/// `--n N`: how many ops the timed loop runs (the harness calibrates it per lane).
fn arg_n(dflt: i64) -> i64 {
    let args: Vec<String> = std::env::args().collect();
    let mut n = dflt;
    let mut i = 1;
    while i < args.len() {
        if args[i] == "--n" && i + 1 < args.len() {
            n = args[i + 1].parse().unwrap_or(dflt);
            i += 1;
        }
        i += 1;
    }
    n.max(2)
}

/// One measured routine, in the row format every lane prints (bench/README.md).
fn row(name: &str, iters: i64, us: i64, items: i64, result: i64, sink: i64) {
    let ns = us * 1000 / iters;
    let per = (us * 1000) as f64 / (iters * items) as f64;
    println!("routine\titers\tus\tns_op\tpx\tns_px\thash");
    println!("{name}\t{iters}\t{us}\t{ns}\t{items}\t{per:.3}\t{result:x}");
    println!("time: {}ms sink={}", us / 1000, black_box(sink));
}

fn main() {
    let n = arg_n(20);
    let count: i64 = 5_000_000;
    let t0 = Instant::now();
    let mut sink: i64 = 0;
    for r in 0..n {
        sink = sink.wrapping_add(mix(black_box(count), black_box(r)));
    }
    let us = t0.elapsed().as_micros() as i64;
    row("sum_loop", n, us, count, mix(count, 0), sink);
}
